contracts/vaults/STETHVault/base/RibbonVault.sol

```diff
@@ -0,0 +1,685 @@
+        uint128 unredeemedShares =
+            depositReceipt.getSharesFromReceipt(
+                currentRound,
+                roundPricePerShare[depositReceipt.round],
+                vaultParams.decimals
+            );
+
+        uint256 depositAmount = amount;
+        // If we have a pending deposit in the current round, we add on to the pending deposit
+        if (currentRound == depositReceipt.round) {
+            uint256 newAmount = uint256(depositReceipt.amount).add(amount);
+            depositAmount = newAmount;
+        }
+
+        ShareMath.assertUint104(depositAmount);
+
+        depositReceipts[creditor] = Vault.DepositReceipt({
+            round: uint16(currentRound),
+            amount: uint104(depositAmount),
+            unredeemedShares: unredeemedShares
+        });
+
+        vaultState.totalPending = uint128(
+            uint256(vaultState.totalPending).add(amount)
+        );
+    }
+
+    /**
+     * @notice Initiates a withdrawal that can be processed once the round completes
+     * @param shares is the number of shares to withdraw
+     */
+    function initiateWithdraw(uint128 shares) external nonReentrant {
+        require(shares > 0, "!shares");
+
+        // We do a max redeem before initiating a withdrawal
+        // But we check if they must first have unredeemed shares
+        if (
+            depositReceipts[msg.sender].amount > 0 ||
+            depositReceipts[msg.sender].unredeemedShares > 0
+        ) {
+            _redeem(0, true);
+        }
+
+        // This caches the `round` variable used in shareBalances
+        uint256 currentRound = vaultState.round;
+        Vault.Withdrawal storage withdrawal = withdrawals[msg.sender];
+
+        bool topup = withdrawal.round == currentRound;
+
+        emit InitiateWithdraw(msg.sender, shares, currentRound);
+
+        uint256 withdrawalShares = uint256(withdrawal.shares);
+
+        if (topup) {
+            uint256 increasedShares = withdrawalShares.add(shares);
+            ShareMath.assertUint128(increasedShares);
+            withdrawals[msg.sender].shares = uint128(increasedShares);
+        } else if (withdrawalShares == 0) {
+            withdrawals[msg.sender].shares = shares;
+            withdrawals[msg.sender].round = uint16(currentRound);
+        } else {
+            // If we have an old withdrawal, we revert
+            // The user has to process the withdrawal
+            revert("Existing withdraw");
+        }
+
+        vaultState.queuedWithdrawShares = uint128(
+            uint256(vaultState.queuedWithdrawShares).add(shares)
+        );
+
+        _transfer(msg.sender, address(this), shares);
+    }
+
+    /**
+     * @notice Completes a scheduled withdrawal from a past round. Uses finalized pps for the round
+     * @param minETHOut is the min amount of `asset` to recieve for the swapped amount of steth in crv pool
+     */
+    function completeWithdraw(uint256 minETHOut) external nonReentrant {
```

Comment:
just a note for when jhh does frontend - minETHOut will need to do extensive math:

1. [shareToUnderlying](https://github.com/ribbon-finance/ribbon-v2/pull/60/files#diff-5020f0dced6f2a6457e6f7cc1592eb79cf21080973d4fab3500f747feab91deaR369)
2. [getWstETHByStETH](https://github.com/ribbon-finance/ribbon-v2/pull/60/files#diff-2bb6f455514e2f789d18593144b618eac195c36fec6d89dd1200da256764c48dR589)
3. get_dy(1, 0, ABOVE_CALCULATION) * slippage / 100